### ISSUE TITLE
Allow empty "required" array

### ISSUE THREAD

**Issue by gibson042** [json-schema-org/json-schema-spec#69]
(copied from https://github.com/json-schema/json-schema/issues/232 )

Draft v4 specifies that a `required` array "MUST have at least one element", which makes it impossible to explicitly indicate that all properties are optional (and therefore also impossible to distinguish intentional optionality from accidental omission of `required`). Empty values (`"required": []`) should be permitted.

**Comment by erosb** [comment]
> which makes it impossible to explicitly indicate that all properties are optional

I don't think it is a problem. Being able to express the same thing in multiple ways (an explicit and an implicit notation) is not a goal of the specification.

**Comment by gibson042** [comment]
Then why is there an `integer` type, to point out just one from a _multitude_ of counterexamples? And why not make `required` required, so author intention is always clear?

Put another way, the default value of `minItems` is 0. Why deviate from that default for `required`?

**Comment by erosb** [comment]
> And why not make required required, so author intention is always clear?
That should be a different proposal.

**Comment by gibson042** [comment]
I was pointing out the inconsistency in claiming that JSON Schema is good about preventing multiple ways of expressing the same thing, not seriously suggesting requiring `required`. _This_ (allowing `required: []`) is the serious proposal, because an empty value conveys information that is not possible if the property is absent (and in fact I've begun defining schemas with `_required: []` as a workaround).

**Comment by awwright** [comment]
I would personally try to leave out a keyword if it's not going to have any effect on validation (since keywords generally add constraints to an instance), but not being overly pedantic about these sorts of things seems pretty reasonable.

There might be a good reason to want to leave in a keyword even if it's not going to do anything.

**Comment by handrews** [comment]
I agree with @awwright about trying to leave it out, but as long as `required` doesn't become required (no keyword should be required, as {} should always be valid and validate everything) I like the idea of allowing `"required": []`

I think mostly we allow putting in the default explicitly, so `required` is currently an exception to that pattern.

**Comment by Relequestual** [comment]
If it says in the spec, that if `required` is not included, then all are optional, then I don't see that there's a problem.

**Comment by epoberezkin** [comment]
No objection here to allowing `required: []`, it's backward compatible :)...

**Comment by awwright** [comment]
Er, as I mentioned in #112, it turns out "dependencies" has this behavior too.

I'd like to just modify "stringArray" so both "dependencies" properties and "required" allows multiple items. They're both used for the same purpose, we should probably modify this behavior together.

**Comment by handrews** [comment]
@gibson042 with #112 now merged (thanks, @awwright !) I think this is now fixed.  If you agree could you please close?

**Comment by epoberezkin** [comment]
In parallel with other keywords in I-D (minLength, etc.), it can be said that `{"required":[]}` is equivalent to `{}`, so `[]` is default for required.

**Comment by awwright** [comment]
Good point, I'll reopen this briefly so I don't forget

**Comment by awwright** [comment]
Addressed in #171

**Comment by handrews** [comment]
@awwright I already included this fix in #168 with all the other meta-schema stuff.
